Markdown Editor

AaronGullickson opened the pull request:
This PR replaces all of the simple JTextArea user entry boxes for descriptions (e.g. personnel biographies, unit histories, scenario after-action reports) with a custom MarkdownEditorPanel that includes buttons for rich text markup and preview. 

Here is an example when the user pops up the "edit biography" for personnel:

<img width="407" alt="Screen Shot 2019-10-04 at 8 36 02 AM" src="https://user-images.githubusercontent.com/16107446/66220725-5b18ce00-e682-11e9-8810-d2d547c9b0fa.png">

Here is a look at the preview tab after some markup has been applied:

<img width="409" alt="Screen Shot 2019-10-04 at 8 42 50 AM" src="https://user-images.githubusercontent.com/16107446/66221026-fd38b600-e682-11e9-94af-10b6e793327c.png">

Here is another example of the panel embedded in a larger dialog:

<img width="420" alt="Screen Shot 2019-10-04 at 8 36 42 AM" src="https://user-images.githubusercontent.com/16107446/66220789-826f9b00-e682-11e9-91f1-e61b1b72f7a5.png">


I modeled it somewhat after the GH issues editor (which I am typing on right now!) and it behaves similarly. It does not give users all of the possible markup available with CommonMark markdown, but does give them most of the basics as well as further information with a link (in the question button). The buttons are hopefully self-explanatory. There are also hotkeys for bold and italicizing.

EDIT - I forgot to mention that the icons come from [icons8.com](https://icons8.com/) and are free to use with acknowledgements which I have added.

NickAragua commented on MekHQ/src/mekhq/gui/dialog/MarkdownEditorDialog.java:
```diff
@@ -0,0 +1,108 @@
+/*
+ * TextAreaDialog.java
```
Needs MekHQ header, meaningful comment for the class, preferably meaningful javadoc comments for the public methods as well.

AaronGullickson replied:
Added javadocs in 0a3caacf0.

NickAragua commented on MekHQ/src/mekhq/gui/utilities/MarkdownEditorPanel.java:
```diff
@@ -0,0 +1,260 @@
+package mekhq.gui.utilities;
```
Needs file header, class comment, public method comments.

AaronGullickson replied:
Added javadocs in b74eda9 and header in 7f7b871

AaronGullickson commented:
Doh, I was planning to add the documentation later and then spaced it. Will do soon.

sixlettervariables commented on MekHQ/src/mekhq/gui/utilities/MarkdownEditorPanel.java:
```diff
@@ -0,0 +1,260 @@
+                    });
+                }
+            }
+        });
```
Indentation is a bit off.

AaronGullickson replied:
fixed indentation in b74eda9

AaronGullickson commented:
@sixlettervariables @NickAragua I think I addressed all of the points above, but there is still one issue with the key bindings for bold and italics. 

```java
editor.getInputMap().put(KeyStroke.getKeyStroke(KeyEvent.VK_B, KeyEvent.CTRL_DOWN_MASK), "bold");
editor.getActionMap().put("bold", new AbstractAction() {
    public void actionPerformed(ActionEvent e) {
                boldText();
    }
});
```

The second line prompts a warning about serialization. On Slack, it was suggested to instead use:

```java
editor.getActionMap().put("bold", e -> boldText());
```

but this gives me errors. 

There is also the issue of licensing for the icons  that I am looking into.

AaronGullickson commented:
Ok with the latest commit, I have changed the icons. Most of the icons come from the [Google Material Design Icons](https://www.iconfinder.com/iconsets/google-material-design-icons). License information is at the bottom of that link. It gives two different Creative Commons licenses but I think the most current one listed is [Creative Commons ShareAlike International 4.0](https://creativecommons.org/licenses/by-sa/4.0/). I have put the following into the contributors page which I think addresses the requirements of the license:

```
  Icons for MarkdownEditorPanel
    Bold, italic, ordered list, unordered list, horizontal rule, and question mark icons provided
    by Google in Google Material Design Icons and distributed under the Creative Commons ShareAlike 4.0 
    International License (https://creativecommons.org/licenses/by-sa/4.0/).
```

The header icon is the only case where I couldn't find something from this icon set, so I used a header icon from [Font Awesome](https://www.iconfinder.com/iconsets/font-awesome) instead. This one says specifically that:

```
Attribution is no longer required as of Font Awesome 3.0, but much appreciated:
Font Awesome by Dave Gandy - http://fontawesome.io
```

So I have added that line to the contributors form as well.
